scvi/inference/total_inference.py

```diff
@@ -476,6 +536,175 @@ def get_normalized_denoised_expression(
+                    label=label,
+                    n_samples=n_samples,
+                    transform_batch=b,
+                )
+                py_ = outputs["py_"]
+                pi = 1 / (1 + torch.exp(-py_["mixing"]))
+                protein_rate += py_["rate_fore"] * (1 - pi) + py_["rate_back"] * pi
+            protein_rate /= len(transform_batch)
+            rate_list_pro.append(protein_rate.cpu())
+
+        if n_samples > 1:
+            # concatenate along batch dimension -> result shape = (samples, cells, features)
+            rate_list_pro = torch.cat(rate_list_pro, dim=1)
+            # (cells, features, samples)
+            rate_list_pro = rate_list_pro.permute(1, 2, 0)
+        else:
+            rate_list_pro = torch.cat(rate_list_pro, dim=0)
+
+        if give_mean is True and n_samples > 1:
+            rate_list_pro = torch.mean(rate_list_pro, dim=-1)
+
+        rate_list_pro = rate_list_pro.cpu().numpy()
+
+        return rate_list_pro
+
+    @torch.no_grad()
+    def generate_denoised_samples(
+        self,
+        n_samples: int = 25,
+        batch_size: int = 64,
+        rna_size_factor: int = 1,
+        transform_batch: Optional[int] = None,
+    ):
+        """ Return samples from an adjusted posterior predictive. Proteins are concatenated to genes.
+        :param n_samples: How may samples per cell
+        :param batch_size: Mini-batch size for sampling. Lower means less GPU memory footprint
+        :rna_size_factor: size factor for RNA prior to sampling gamma distribution
+        :transform_batch: int of which batch to condition on for all cells
+        :return:
+        """
+        posterior_list = []
+        for tensors in self.update({"batch_size": batch_size}):
+            x, _, _, batch_index, labels, y = tensors
+            with torch.no_grad():
+                outputs = self.model.inference(
+                    x,
+                    y,
+                    batch_index=batch_index,
+                    label=labels,
+                    n_samples=n_samples,
+                    transform_batch=transform_batch,
+                )
+            px_ = outputs["px_"]
+            py_ = outputs["py_"]
+
+            pi = 1 / (1 + torch.exp(-py_["mixing"]))
+            mixing_sample = Bernoulli(pi).sample()
+            protein_rate = py_["rate_fore"]
+            rate = torch.cat((rna_size_factor * px_["scale"], protein_rate), dim=-1)
+            if len(px_["r"].size()) == 2:
+                px_dispersion = px_["r"]
+            else:
+                px_dispersion = torch.ones_like(x) * px_["r"]
+            if len(py_["r"].size()) == 2:
+                py_dispersion = py_["r"]
+            else:
+                py_dispersion = torch.ones_like(y) * py_["r"]
+
+            dispersion = torch.cat((px_dispersion, py_dispersion), dim=-1)
+
+            # This gamma is really l*w using scVI manuscript notation
+            p = rate / (rate + dispersion)
```

Comment:
this appears now multiple times in the codebase. Should we have some basic code for sampling from NB ? These numpy comments and parameterization of the NB are always heavy

Reply:
we should really write our own custom NB distribution, which subclasses pytorch objects.

Reply:
It's not really the place to do this here. We could also write our own Gamma distribution that uses our parameterization.

Reply:
yes agreed. We should write a github issue